syntax = "proto3";

package geospatial;

// ========== Project Management Messages ==========

// Helper messages for complex data structures
message PreviewRow {
  repeated string values = 1;
}

message DataRow {
  map<string, string> fields = 1;
}

// Project data structure
message Project {
  string id = 1;
  string name = 2;
  string description = 3;
  int64 created_at = 4;  // Unix timestamp
  int64 updated_at = 5;  // Unix timestamp
}

// File data structure with dataset type
message File {
  string id = 1;
  string project_id = 2;
  string name = 3;
  DatasetType dataset_type = 4;
  string original_filename = 5;
  int64 file_size = 6;
  int64 created_at = 7;  // Unix timestamp
}

// Dataset type enumeration
enum DatasetType {
  DATASET_TYPE_UNSPECIFIED = 0;
  DATASET_TYPE_SAMPLE = 1;
  DATASET_TYPE_DRILL_HOLES = 2;
  DATASET_TYPE_BLOCK = 3;
}

// Column type enumeration for CSV processing
enum ColumnType {
  COLUMN_TYPE_UNSPECIFIED = 0;
  COLUMN_TYPE_NUMERIC = 1;
  COLUMN_TYPE_CATEGORICAL = 2;
  COLUMN_TYPE_UNUSED = 3;
}

// Column mapping for CSV processing
message ColumnMapping {
  string column_name = 1;
  ColumnType column_type = 2;
  string mapped_field = 3;  // "x", "y", "z", or custom field name
  bool is_coordinate = 4;   // true for X, Y, Z coordinates
}

// Dataset data structure
message Dataset {
  string id = 1;
  string file_id = 2;
  int32 total_rows = 3;
  int32 current_page = 4;
  repeated ColumnMapping column_mappings = 5;
  int64 created_at = 6;  // Unix timestamp
}

// ========== Request/Response Messages ==========

// Project CRUD operations
message CreateProjectRequest {
  string name = 1;
  string description = 2;
}

message CreateProjectResponse {
  Project project = 1;
  bool success = 2;
  string error_message = 3;
}

message GetProjectsRequest {
  int32 limit = 1;
  int32 offset = 2;
}

message GetProjectsResponse {
  repeated Project projects = 1;
  int32 total_count = 2;
}

message GetProjectRequest {
  string project_id = 1;
}

message GetProjectResponse {
  Project project = 1;
  bool success = 2;
  string error_message = 3;
}

message UpdateProjectRequest {
  string project_id = 1;
  string name = 2;
  string description = 3;
}

message UpdateProjectResponse {
  Project project = 1;
  bool success = 2;
  string error_message = 3;
}

message DeleteProjectRequest {
  string project_id = 1;
}

message DeleteProjectResponse {
  bool success = 1;
  string error_message = 2;
}

// File operations
message CreateFileRequest {
  string project_id = 1;
  string name = 2;
  DatasetType dataset_type = 3;
  string original_filename = 4;
  bytes file_content = 5;
}

message CreateFileResponse {
  File file = 1;
  bool success = 2;
  string error_message = 3;
}

message GetProjectFilesRequest {
  string project_id = 1;
}

message GetProjectFilesResponse {
  repeated File files = 1;
}

message DeleteFileRequest {
  string file_id = 1;
}

message DeleteFileResponse {
  bool success = 1;
  string error_message = 2;
}

// Update file metadata
message UpdateFileRequest {
  string file_id = 1;
  string name = 2;
}

message UpdateFileResponse {
  File file = 1;
  bool success = 2;
  string error_message = 3;
}

// Rename file columns in DuckDB
message RenameFileColumnRequest {
  string file_id = 1;
  map<string, string> column_renames = 2;  // old_name -> new_name
}

message RenameFileColumnResponse {
  repeated string renamed_columns = 1;  // List of new column names
  bool success = 2;
  string error_message = 3;
}

// Get file statistics
message GetFileStatisticsRequest {
  string file_id = 1;
  repeated string columns = 2;  // Optional: specific columns to get stats for
}

message ColumnStatistics {
  string column_name = 1;
  string data_type = 2;  // "numeric" or "categorical"
  int32 count = 3;
  int32 null_count = 4;
  int32 unique_count = 5;

  // For numeric columns
  optional double mean = 6;
  optional double std = 7;
  optional double min = 8;
  optional double q25 = 9;
  optional double q50 = 10;  // median
  optional double q75 = 11;
  optional double max = 12;

  // For categorical columns
  repeated string top_values = 13;
  repeated int32 top_counts = 14;
}

message GetFileStatisticsResponse {
  repeated ColumnStatistics statistics = 1;
  bool success = 2;
  string error_message = 3;
}

// ========== Data Manipulation Operations ==========

// Replace file data (data cleaning)
message DataReplacement {
  string from_value = 1;
  string to_value = 2;
}

message ReplaceFileDataRequest {
  string file_id = 1;
  repeated DataReplacement replacements = 2;
  repeated string columns = 3;  // Optional: specific columns to apply replacements
}

message ReplaceFileDataResponse {
  int32 rows_affected = 1;
  bool success = 2;
  string error_message = 3;
}

// Search/filter file data with pagination
message SearchFileDataRequest {
  string file_id = 1;
  string query = 2;  // Search query/filter condition
  int32 limit = 3;   // Max results to return
  int32 offset = 4;  // Pagination offset
}

message SearchFileDataResponse {
  string file_id = 1;
  int32 total_rows = 2;
  int32 current_page = 3;
  repeated DataRow data = 4;
  bool success = 5;
  string error_message = 6;
}

// Filter file data with optional new file creation
message FilterFileDataRequest {
  string file_id = 1;
  string column = 2;
  string operation = 3;  // "=", "!=", ">", "<", ">=", "<=", "LIKE"
  string value = 4;
  bool create_new_file = 5;
  string new_file_name = 6;  // Required if create_new_file=true
}

message FilterFileDataResponse {
  string file_id = 1;  // Original or new file_id
  int32 total_rows = 2;
  bool success = 3;
  string error_message = 4;
}

// Delete specific points/rows from file
message DeleteFilePointsRequest {
  string file_id = 1;
  repeated int32 row_indices = 2;  // Row indices to delete
}

message DeleteFilePointsResponse {
  int32 rows_deleted = 1;
  int32 rows_remaining = 2;
  bool success = 3;
  string error_message = 4;
}

// Add a filtered column (non-destructive filtering)
message AddFilteredColumnRequest {
  string file_id = 1;
  string new_column_name = 2;       // Name for the new column
  string source_column = 3;          // Column to filter on
  string operation = 4;              // "=", "!=", ">", "<", ">=", "<=", "LIKE"
  string value = 5;                  // Value to compare against
}

message AddFilteredColumnResponse {
  string new_column_name = 1;
  int32 rows_with_values = 2;        // Number of rows matching the filter
  int32 rows_with_null = 3;          // Number of rows with NULL (non-matching)
  bool success = 4;
  string error_message = 5;
}

// ========== Advanced Column Operations ==========

// Add new columns to file
message NewColumn {
  string column_name = 1;
  repeated string values = 2;  // Must match row count
}

message AddFileColumnsRequest {
  string file_id = 1;
  repeated NewColumn new_columns = 2;
}

message AddFileColumnsResponse {
  repeated string added_columns = 1;
  bool success = 2;
  string error_message = 3;
}

// Duplicate existing columns
message ColumnDuplicate {
  string source_column = 1;
  string new_column_name = 2;  // Optional - if empty, auto-generate as source_column_copy
}

message DuplicateFileColumnsRequest {
  string file_id = 1;
  repeated ColumnDuplicate columns = 2;
}

message DuplicateFileColumnsResponse {
  repeated string duplicated_columns = 1;  // New column names created
  bool success = 2;
  string error_message = 3;
}

// ========== Dataset Merging ==========

enum MergeMode {
  MERGE_MODE_UNSPECIFIED = 0;
  MERGE_MODE_BY_ROWS = 1;      // UNION ALL - append rows
  MERGE_MODE_BY_COLUMNS = 2;   // JOIN - add columns
}

message MergeDatasetsRequest {
  string first_dataset_id = 1;
  string second_dataset_id = 2;
  MergeMode mode = 3;

  // Only for BY_COLUMNS mode
  repeated string exclude_columns_first = 4;   // Columns to exclude from first dataset
  repeated string exclude_columns_second = 5;  // Columns to exclude from second dataset

  string output_file = 6;  // Optional: name for merged file
}

message MergeDatasetsResponse {
  string dataset_id = 1;         // ID of resulting dataset
  int32 rows_merged = 2;
  int32 columns_merged = 3;
  repeated string warnings = 4;
  bool success = 5;
  string error_message = 6;
}

// Enhanced CSV analysis with column type detection
message AnalyzeCsvForProjectRequest {
  string file_id = 1;
}

message AnalyzeCsvForProjectResponse {
  repeated string headers = 1;
  repeated PreviewRow preview_rows = 2;
  repeated ColumnType suggested_types = 3;  // AI-suggested column types
  map<string, string> suggested_mappings = 4;  // column_name -> suggested field (x/y/z)
  int32 total_rows = 5;
  bool success = 6;
  string error_message = 7;
}

// Dataset processing with column mappings
message ProcessDatasetRequest {
  string file_id = 1;
  repeated ColumnMapping column_mappings = 2;
}

message ProcessDatasetResponse {
  Dataset dataset = 1;
  bool success = 2;
  string error_message = 3;
  int32 processed_rows = 4;
}

// Data boundaries for visualization scaling
message DataBoundaries {
  string column_name = 1;
  double min_value = 2;
  double max_value = 3;
  int32 valid_count = 4;  // Number of valid (non-null, numeric) values
}

// Pre-computed statistics for efficient frontend visualization

// Histogram data for bar charts
message HistogramData {
  repeated string bin_ranges = 1;      // Human-readable ranges: ["0.00 - 10.00", "10.00 - 20.00", ...]
  repeated int32 bin_counts = 2;       // Count of values in each bin
  repeated double bin_edges = 3;       // Numeric bin edges for precise rendering
  int32 num_bins = 4;                  // Number of bins used
  double min_value = 5;                // Minimum value in data
  double max_value = 6;                // Maximum value in data
  int32 total_count = 7;               // Total number of values
}

// Box plot statistical data
message BoxPlotData {
  string column_name = 1;
  double min = 2;                      // Minimum value (excluding outliers)
  double q1 = 3;                       // First quartile (25th percentile)
  double median = 4;                   // Median (50th percentile)
  double q3 = 5;                       // Third quartile (75th percentile)
  double max = 6;                      // Maximum value (excluding outliers)
  double mean = 7;                     // Mean/average value
  repeated double outliers = 8;        // Outlier values
  double lower_fence = 9;              // Lower fence (Q1 - 1.5*IQR)
  double upper_fence = 10;             // Upper fence (Q3 + 1.5*IQR)
  double iqr = 11;                     // Interquartile range (Q3 - Q1)
  int32 total_count = 12;              // Total number of values
}

// Heatmap cell data
message HeatmapCell {
  int32 x_index = 1;                   // X bin index
  int32 y_index = 2;                   // Y bin index
  double avg_value = 3;                // Average value in this cell
  int32 count = 4;                     // Number of points in this cell
}

// Heatmap aggregated data
message HeatmapData {
  repeated HeatmapCell cells = 1;      // Grid cells with aggregated values
  int32 grid_size_x = 2;               // Number of bins in X direction
  int32 grid_size_y = 3;               // Number of bins in Y direction
  double min_value = 4;                // Minimum aggregated value
  double max_value = 5;                // Maximum aggregated value
  double x_bin_size = 6;               // Width of each X bin
  double y_bin_size = 7;               // Height of each Y bin
  double min_x = 8;                    // Minimum X coordinate
  double max_x = 9;                    // Maximum X coordinate
  double min_y = 10;                   // Minimum Y coordinate
  double max_y = 11;                   // Maximum Y coordinate
  string x_column = 12;                // X column name
  string y_column = 13;                // Y column name
  string value_column = 14;            // Value column name
}

// Get dataset data with pagination
message GetDatasetDataRequest {
  string dataset_id = 1;
  repeated string columns = 2;

  // Optional filtering and visualization parameters
  optional string shape = 3;           // e.g. "circle" | "rect" | "triangle" | ...
  optional string color = 4;           // color name or hex
  optional string function = 5;        // e.g. "IDW" | "NONE" | "log" (default: "NONE")
  repeated double bounding_box = 6;    // [x1, x2, y1, y2] for 2D or [x1, x2, y1, y2, z1, z2] for 3D
}

message GetDatasetDataResponse {
  bytes binary_data = 1;  // Dynamic row data
  int32 data_length = 2;
  int32 total_count = 3;
  repeated DataBoundaries data_boundaries = 4;

  // Pre-computed statistics for visualization (computed on backend)
  map<string, HistogramData> histograms = 5;     // Key: column_name, Value: histogram data
  repeated BoxPlotData box_plots = 6;             // Box plot data for all numeric columns
  optional HeatmapData heatmap = 7;               // Heatmap aggregation (if x,y,value columns provided)
}

// Note: DataBoundaries are now included directly in GetDatasetDataResponse
// No separate boundaries method needed

// Get datasets for a project
message GetProjectDatasetsRequest {
  string project_id = 1;
}

message DatasetInfo {
  string id = 1;
  string file_id = 2;
  string file_name = 3;
  int32 dataset_type = 4;
  string original_filename = 5;
  int32 total_rows = 6;
  int64 created_at = 7;
  repeated ColumnMapping column_mappings = 8;
}

message GetProjectDatasetsResponse {
  repeated DatasetInfo datasets = 1;
}

// Get paginated table data for dataset (for displaying in tables/grids)
message GetDatasetTableDataRequest {
  string dataset_id = 1;
  int32 limit = 2;           // Number of rows to return (e.g., 1000)
  int32 offset = 3;          // Number of rows to skip (for pagination)
  repeated string columns = 4;  // Optional: specific columns to fetch (empty = all columns)
}

message TableRow {
  map<string, double> values = 1;  // Column name -> value mapping
}

message GetDatasetTableDataResponse {
  repeated TableRow rows = 1;
  int32 total_rows = 2;           // Total number of rows in dataset
  repeated string column_names = 3; // Column names in order
  bool success = 4;
  string error_message = 5;
}

// Delete dataset
message DeleteDatasetRequest {
  string dataset_id = 1;
}

message DeleteDatasetResponse {
  bool success = 1;
  string error_message = 2;
  int32 rows_deleted = 3;  // Number of data rows deleted
  double delete_time = 4;  // Time taken in seconds
}